Client/Frontend/NotificationSurface/NotificationSurfaceManager.swift

```diff
@@ -81,7 +79,38 @@ class NotificationSurfaceManager: NotificationSurfaceDelegate {
         }
     }
 
+    func didDismissNotification(_ userInfo: [AnyHashable: Any]) {
+        guard let messageId = userInfo[Constant.messageIdKey] as? String,
+              let message = messagingManager.messageForId(messageId, surface: notificationSurfaceID)
+        else { return }
+
+        messagingManager.onMessageDismissed(message)
+    }
+
     // MARK: - Private
+    private func scheduleNotification(message: GleanPlumbMessage, notificationId: String) {
+        let notificationCategory = UNNotificationCategory(
+            identifier: Constant.notificationCategoryId,
+            actions: [],
+            intentIdentifiers: [],
+            options: .customDismissAction)
+
+        UNUserNotificationCenter.current().setNotificationCategories([notificationCategory])
+
+        let userInfo = [Constant.messageIdKey: message.id]
+        notificationManager.schedule(title: message.data.title ?? "",
+                                     body: message.data.text,
+                                     id: notificationId,
+                                     userInfo: userInfo,
+                                     categoryIdentifier: Constant.notificationCategoryId,
+                                     interval: TimeInterval(Constant.messageDelay),
+                                     repeats: false)
+
+        // Schedule notification telemetry for when notification gets displayed
+        DispatchQueue.main.asyncAfter(deadline: .now() + Constant.messageDelay) { [weak self] in
+            self?.didDisplayMessage(message)
```

Comment:
As far as I understand it, `didDisplayMessage` just does bookkeeping, and doesn't touch the UI, so why does this happen a) asynchronously b) on the main thread?

Reply:
As the notification is getting scheduled 3 seconds after the background task is run I wanted to send `didDisplayMessage` at the same time as the message will be displayed to the user. Scheduling can only be done async as we are otherwise blocking the main thread. I had originally had it executed on the main thread because I didn't know what we were doing was thread safe. I looked into it now and we should be good to execute on a background thread so I changed it.